### PR TITLE
feat: separate flags for request/response E2E checksum and enable request checksum by default

### PR DESCRIPTION
**Pull request by nimf**
This PR replaces the `GOOGLE_CLOUD_DATASTORE_HTTP_ENABLE_E2E_CHECKSUM` flag with two:

- `GOOGLE_CLOUD_DATASTORE_HTTP_ENABLE_E2E_REQUEST_CHECKSUM` - attach payload checksum header to requests (enabled by default)
- `GOOGLE_CLOUD_DATASTORE_HTTP_ENABLE_E2E_RESPONSE_CHECKSUM` - verify response checksum (if extsts in response headers) against response payload and raise IOException if it doesn't match (disabled by default)

And enables the E2E request checksum by default.
This is safe because nobody is using the old flag and we don't reject calls with invalid checksum on the server side yet. But this will provide us the analytics of number of requests with valid/invalid checksums to find edge cases and better prepare for E2E checksum rollout.

The tests for `RemoteRpc` using request and response checksum was also reworked.

Previously the test called the protected `setHeaders` but it does not verify that `RemoteRpc` will call it when making a call. I. e., if in the future we change `RemoteRpc`'s `call` method and remake adding the headers without using the `setHeaders` method then the tests will remain green while necessary headers may not be added.

Instead of calling `setHeaders` from the test itself, we pass the expected headers to the `MyHttpTransport` and make sure the required headers are in place when `MyLowLevelHttpRequest` is doing its `execute` call.

It is easy to convert the remaining test (not related to E2E checksum) that uses `setHeaders` to the same approach, then we can make `setHeaders` private. I can do it in a separate PR if you like.

Another change in the test is that now we verify that checksum in the header is generated correctly (not just checking the length of the checksum string) and verify that the response checksum gets validated.

### CODE REVIEW

**Review by jainsahab — DISMISSED** [review]
Just one small nit, otherwise LGTM.

File: datastore-v1-proto-client/src/test/java/com/google/datastore/v1/client/RemoteRpcTest.java [hunk 1]
```diff
@@ -289,24 +360,103 @@ protected LowLevelHttpRequest buildRequest(String method, String url) throws IOE
     }
   }
 
+  private static class MyHeader {
+    private final String key;
+    private final String value;
+    private final boolean ignoreValue;
+    private boolean mustExist;
+
+    public static MyHeader AnyValue(String key) {
```
[the remark this answers is not in the document]
Comment (nimf) [reply datastore-v1-proto-client/src/test/java/com/google/datastore/v1/client/RemoteRpcTest.java]: oops, jumping between Java/Go has its consequences, haha. Done.

File: datastore-v1-proto-client/src/main/java/com/google/datastore/v1/client/RemoteRpc.java [hunk 2]
```diff
@@ -53,9 +53,18 @@ class RemoteRpc {
   private final HttpRequestInitializer initializer;
   private final String url;
   private final AtomicInteger rpcCount = new AtomicInteger(0);
-  // Not final - so it can be set/reset in Unittests
-  private static boolean enableE2EChecksum =
-      Boolean.parseBoolean(System.getenv("GOOGLE_CLOUD_DATASTORE_HTTP_ENABLE_E2E_CHECKSUM"));
```
[the remark this answers is not in the document]
Comment (nimf) [reply datastore-v1-proto-client/src/main/java/com/google/datastore/v1/client/RemoteRpc.java:58]: If a user has the old env var set then it will be ignored and request checksum will remain enabled (as it is enabled by default now) and response checksum will become disabled.

I can provide a transition so that if the old env var is set then both checksums are enabled. But I specifically decided not to do that because of two reasons:

1. No one has the old env var set (I have an ESF metric for that).
2. We don't want to encourage enabling response checksum verification yet until we have a metric for success/error rates and figured out all edge-cases.

[the remark this answers is not in the document]
Comment (kolea2) [reply datastore-v1-proto-client/src/main/java/com/google/datastore/v1/client/RemoteRpc.java:58]: got it, thanks! I think as an extra precaution, we should still check if this env variable is set and simply log a warning to the user that they should migrate/update their code.

[the remark this answers is not in the document]
Comment (nimf) [reply datastore-v1-proto-client/src/main/java/com/google/datastore/v1/client/RemoteRpc.java:58]: okay, sounds good. Added the handling of previous env var (it will still work -- enabling/disabling both request and response checksums if only the old env var is provided) and a warning. Is the wording for the warning okay?

[the remark this answers is not in the document]
Comment (kolea2) [reply datastore-v1-proto-client/src/main/java/com/google/datastore/v1/client/RemoteRpc.java:58]: sorry if I was unclear - I think it's fine to *only* warn that the env variable is set, and not actually use it. Feel free to change the logic and warning message accordingly (the wording lgtm)

[the remark this answers is not in the document]
Comment (nimf) [reply datastore-v1-proto-client/src/main/java/com/google/datastore/v1/client/RemoteRpc.java:58]: no, I decided to keep the behavior of the old env variable so that it won't be a breaking change at all.

**Review by kolea2 — DISMISSED** [review]
